pkg/informer/informer.go

```diff
@@ -530,56 +568,76 @@ func (d *DynamicDiscoverySharedInformerFactory) updateInformers() {
+// includes enough data in each APIResource to support what kcp currently needs (scheduling, placement, quota, GC).
+func gvrsToDiscoveryData(gvrs map[schema.GroupVersionResource]gvrPartialMetadata) discoveryData {
+	ret := discoveryData{
+		apiGroupResources: make([]*restmapper.APIGroupResources, 0),
+		apiResourceList:   make([]*metav1.APIResourceList, 0),
+	}
 
-	for gvr := range gvrs {
-		gv := gvr.GroupVersion()
+	gvResources := make(map[string]map[string][]metav1.APIResource)
 
-		gvResources[gv] = append(gvResources[gv], metav1.APIResource{
-			Name:    gvr.Resource,
-			Group:   gvr.Group,
-			Version: gvr.Version,
+	for gvr, metadata := range gvrs {
+		apiResource := metav1.APIResource{
+			Name:         gvr.Resource,
+			Group:        gvr.Group,
+			Version:      gvr.Version,
+			Kind:         metadata.Names.Kind,
+			SingularName: metadata.Names.Singular,
+			Namespaced:   metadata.Scope == apiextensionsv1.NamespaceScoped,
 			// Everything we're informing on supports these
 			Verbs: []string{"create", "list", "watch", "delete"},
-		})
+		}
+		if gvResources[gvr.Group] == nil {
+			gvResources[gvr.Group] = make(map[string][]metav1.APIResource)
+		}
+		gvResources[gvr.Group][gvr.Version] = append(gvResources[gvr.Group][gvr.Version], apiResource)
 	}
 
-	for gv, resources := range gvResources {
-		sort.Slice(resources, func(i, j int) bool {
-			return resources[i].Name < resources[j].Name
-		})
+	for group, resources := range gvResources {
+		var versions []metav1.GroupVersionForDiscovery
+		versionedResources := make(map[string][]metav1.APIResource)
 
-		discoveryData = append(discoveryData, &metav1.APIResourceList{
-			GroupVersion: gv.String(),
-			APIResources: resources,
-		})
-	}
+		for version, apiResource := range resources {
+			versions = append(versions, metav1.GroupVersionForDiscovery{GroupVersion: group, Version: version})
 
-	sort.Slice(discoveryData, func(i, j int) bool {
-		return discoveryData[i].GroupVersion < discoveryData[j].GroupVersion
-	})
+			sort.Slice(apiResource, func(i, j int) bool {
+				return apiResource[i].Name < apiResource[j].Name
+			})
 
-	return discoveryData
-}
+			versionedResources[version] = apiResource
 
-// DiscoveryData implements resourcequota.NamespacedResourcesFunc and is intended to be used by the quota subsystem.
-func (d *DynamicDiscoverySharedInformerFactory) DiscoveryData() ([]*metav1.APIResourceList, error) {
-	d.informersLock.RLock()
-	defer d.informersLock.RUnlock()
+			apiResourceList := &metav1.APIResourceList{
+				GroupVersion: metav1.GroupVersion{Group: group, Version: version}.String(),
+				APIResources: apiResource,
+			}
 
-	ret := make([]*metav1.APIResourceList, len(d.discoveryData))
-	for i, apiResourceList := range d.discoveryData {
-		ret[i] = apiResourceList.DeepCopy()
+			ret.apiResourceList = append(ret.apiResourceList, apiResourceList)
+		}
+		apiGroup := metav1.APIGroup{
+			Name:     group,
+			Versions: versions,
+			// TODO: PreferredVersion
```

Comment:
TODO when? Why? What's the implication of leaving it out now?

Reply:
Added a comment with more context about it.